### DISCUSSION TITLE
How to apply external pose estimation

### DISCUSSION THREAD

**Discussion by allegorywrite** [bitcraze/discussions#1173] [Q&A]
Hi. [The project to integrate VIO into crazyflie](https://github.com/orgs/bitcraze/discussions/1032) is progressing, and we are ready to send odometry from VIO to crazyflie using [send_extpose](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/api/cflib/crazyflie/extpos/). The frequency of pose estimation required to control crazyflie is 100~500Hz, while the odometry obtained from vio is ~10Hz, so it is necessary to integrate the two well.
When I simply used send_extpose(&kalman filter), I noticed that there was a misalign between the sent frame and the pose estimated inside crazyfllie. This misalign seems to increase as crazyflie moves away from the origin. Why do you think this happen?
![Screenshot from 2024-02-05 19-57-49](https://github.com/bitcraze/discussions/assets/63725656/bfdfc17f-889f-4f60-985d-640386ecf44c)

I understand that if you send odometry to crazyflie that is completely different from the orientation observed by imu (especially in the z-axis direction), the kalman filter will output an orientation that is different from the orientation sent. Also, the coordinate system of crazyflie doesn't seem to be a common setting (pitch is counterclockwise!), so you need to be careful when using pitch, but since send_extpose uses quaternion directly, I believe there is no need to consider it　in this case (The code is shown below just in case)

```c++
void send_pose(const Eigen::Vector3d &vio_t, const Eigen::Quaterniond &vio_q, const std_msgs::Header &header)
{
    motion_capture_tracking_msgs::NamedPoseArray pose_array;
    pose_array.header = header;
    pose_array.header.frame_id = "world";
    motion_capture_tracking_msgs::NamedPose pose;
    pose.name = "drone_" + std::to_string(DRONE_ID) + "_cf";
    pose.pose.position.x = vio_t.x();
    pose.pose.position.y = vio_t.y();
    pose.pose.position.z = vio_t.z();
    pose.pose.orientation.w = vio_q.w();
    pose.pose.orientation.x = vio_q.x();
    pose.pose.orientation.y = vio_q.y();
    pose.pose.orientation.z = vio_q.z();
    pose_array.poses.push_back(pose);
    pose_array.header.stamp = ros::Time::now();
    pub_cf_pose.publish(pose_array);

    static tf::TransformBroadcaster br;
    tf::Transform transform;
    tf::Quaternion q;
    transform.setOrigin(tf::Vector3(vio_t.x(), vio_t.y(), vio_t.z()));
    q.setW(vio_q.w());
    q.setX(vio_q.x());
    q.setY(vio_q.y());
    q.setZ(vio_q.z());
    transform.setRotation(q);
    br.sendTransform(tf::StampedTransform(transform, header.stamp, "world", "send_body"));
}
``` 

Do you know how to completely trust the external pose estimate or how to resolve the issue causing frame misalign? Any tips are welcome.

**Comment by knmcguire** [comment]
Hi!

you are right in the sense that you'll need to send ext_pose in the right [orientation coordinate system](https://www.bitcraze.io/documentation/system/platform/cf2-coordinate-system/) that the crazyflie uses. There seems to be an mismatch... This mocap example would also be good to look at perhaps: https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/mocap/mocap_hl_commander.py. 

So the code that you send, seems to be ROS code and not send_extpose. So it is a bit difficult to see what is going on. Also, it is not clear to see which orientation the VIO estimated pose is using, so i'm not able to see what is going wrong here.

**Reply by allegorywrite** [reply]
The topic is sent from ROS1(code above) to ROS2 via ros1_bridge, and send_extpose is executed using the function below.
```python
def _poses_changed(self, msg):
        """
        Topic update callback to the motion capture lib's
           poses topic to send through the external position
           to the crazyflie 
        """

        poses = msg.poses
        for pose in poses:
            name = pose.name
            x = pose.pose.position.x
            y = pose.pose.position.y
            z = pose.pose.position.z
            quat = pose.pose.orientation
            self.get_logger().info(f"Send extpos {x}, {y}, {z} to {name}")

            if name in self.uri_dict.keys():
                self.get_logger().info(f"name:{name}")
                uri = self.uri_dict[name]
                if isnan(quat.x):
                    self.swarm._cfs[uri].cf.extpos.send_extpos(
                        x, y, z)
                else:
                    self.swarm._cfs[uri].cf.extpos.send_extpose(
                        x, y, z, quat.x, quat.y, quat.z, quat.w)

``` 
I'm not using QoS temporarily, but I don't know if this has any effect.
```python
# self.create_subscription(
            #     NamedPoseArray, "/poses", 
            #     self._poses_changed, qos_profile
            # )
            self.create_subscription(
                NamedPoseArray, "/poses", 
                self._poses_changed, 10
            )
```

**Comment by allegorywrite** [comment]
I could reproduce this problem by simple demo.
```python
#!/usr/bin/env python3

import rclpy
import numpy as np
from rclpy.node import Node
from std_msgs.msg import Header
from geometry_msgs.msg import Pose
from sensor_msgs.msg import Image, Imu

from motion_capture_tracking_interfaces.msg import NamedPoseArray, NamedPose
from rclpy.qos import QoSProfile, QoSReliabilityPolicy, QoSHistoryPolicy
from rclpy.duration import Duration
import cv2
from cv_bridge import CvBridge

import tf_transformations
from tf2_ros import TransformBroadcaster
from geometry_msgs.msg import PoseStamped, TransformStamped

class PosePublisher(Node):
    def __init__(self):
        super().__init__('pose_publisher')
        qos_profile = QoSProfile(
            reliability=QoSReliabilityPolicy.BEST_EFFORT,
            history=QoSHistoryPolicy.KEEP_LAST,
            depth=1,
            deadline=Duration(seconds=0, nanoseconds=1e9 / 100.0)
        )
        self.tfbr = TransformBroadcaster(self)
        self.world_tf_name = "world"
        self.bridge = CvBridge()
        self.publisher = self.create_publisher(NamedPoseArray, '/poses', qos_profile)
        self.imu_subscription = self.create_subscription(Imu, '/imu', self.imu_callback, 10)
        timer_period = 0.1  # seconds
        self.timer = self.create_timer(timer_period, self.timer_callback)
        
    def timer_callback(self):
        cf_name = "drone_0_cf"
        noise = np.random.normal(0, 0.01, 4)
        x = 0.0
        y = 0.0
        z = 1.0
        q = tf_transformations.quaternion_from_euler(0, 0,np.pi/2)
        named_pose_array_msg = NamedPoseArray()
        named_pose_array_msg.header = Header()
        named_pose_array_msg.header.stamp = self.get_clock().now().to_msg()
        named_pose_array_msg.header.frame_id = self.world_tf_name
        pose_cf231 = Pose()
        pose_cf231.position.x = x
        pose_cf231.position.y = y
        pose_cf231.position.z = z
        pose_cf231.orientation.x = q[0]
        pose_cf231.orientation.y = q[1]
        pose_cf231.orientation.z = q[2]
        pose_cf231.orientation.w = q[3]

        named_pose_array_msg.poses = [
            NamedPose(name=cf_name, pose=pose_cf231),
        ]
        
        self.get_logger().info(f'Publishing: {named_pose_array_msg}')
        
        self.publisher.publish(named_pose_array_msg)

        t_base = TransformStamped()
        t_base.header.stamp = self.get_clock().now().to_msg()
        t_base.header.frame_id = self.world_tf_name
        t_base.child_frame_id = cf_name + "_send"
        t_base.transform.translation.x = x
        t_base.transform.translation.y = y
        t_base.transform.translation.z = z
        t_base.transform.rotation.x = q[0]
        t_base.transform.rotation.y = q[1]
        t_base.transform.rotation.z = q[2]
        t_base.transform.rotation.w = q[3]
        self.tfbr.sendTransform(t_base)

    def imu_callback(self, msg):
        self.get_logger().info(f'imu: {msg}')

def main(args=None):
    rclpy.init(args=args)
    pose_publisher = PosePublisher()
    rclpy.spin(pose_publisher) 
    pose_publisher.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()
``` 
If you set the pose to tf_transformations.quaternion_from_euler(0, 0,np.pi/2), the odometry via the Kalman filter will converge to a state different from pi/2.
![Screenshot from 2024-02-08 18-16-28](https://github.com/bitcraze/discussions/assets/63725656/c0206a38-71dd-46f1-8766-0573a61e0dd5)

**Reply by knmcguire** [reply]
Thanks for this information. 

So technically in the EKF, when the orientation is send by mocap, the angle rates are calculated based on the last state and that is the input for the update. some approximations are being made that makes this work somehow: https://github.com/bitcraze/crazyflie-firmware/blob/aefa1d6ed7c8f363d43473e02c53211be427b933/src/modules/src/kalman_core/mm_pose.c#L41-L60

The weird thing is, that I would then assume that the starting orientation does not matter as it is based on rate anyway, but for some reason it does and I haven't been able to figure out why... 

Perhaps for the sake of science, you can maybe check in the space that the VIO's approximation of the orientation is, and try to startup the crazyflie in that same orientation such that it is pointing in the same direction. Changing the start orientation of the crazyflie can also be changed with this script: https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/positioning/initial_position.py
